instance/models/load_balancer.py

```diff
@@ -238,21 +238,30 @@ def reconfigure(self, triggering_instance_id=None):
         The triggering_instance_id indicates the id of the instance reference that initiated the
         reconfiguration of the load balancer.
         """
-        # This is subject to a race condition, but it doesn't matter:
-        # if two processes increased the version concurrently, the LB would be marked dirty
-        # either way, and only one could successfully grab the lock, while the other would get
-        # kicked out after the blocking timeout.
-        self.configuration_version += 1
-        self.save()
-
-        # Memorize the configuration version, in case new threads change it.
+        # We need to use an F expression here.  The problem is not other processes trying to
+        # increase this counter concurrently – that wouldn't matter, since we don't care whether
+        # we increase this counter by one or by two, since both marks the LB as dirty.  However, if
+        # another process is making a completely unrelated change to the LB object we might lose
+        # the increment altogether.
+        LoadBalancingServer.objects.filter(pk=self.pk).update(
+            configuration_version=models.F("configuration_version") + 1
+        )
+        self.refresh_from_db()
+
         try:
             with self._configuration_lock(blocking=False):
+                # Memorize the configuration version, in case new threads change it.
                 candidate_configuration_version = self.configuration_version
+
                 self.logger.info("Reconfiguring load-balancing server %s", self.domain)
                 self.run_playbook(self.get_ansible_vars(triggering_instance_id))
-                self.deployed_configuration_version = candidate_configuration_version
-                self.save()
+
+                with transaction.atomic():
+                    # Reload to pick up the potentially changed configuration_version
+                    self.refresh_from_db()
+                    self.deployed_configuration_version = candidate_configuration_version
+                    self.save()
```

Comment:
Can you just replace this whole `with` statement with another `LoadBalancingServer.objects.filter(pk=self.pk).update(deployed_configuration_version= candidate_configuration_version)` to be sure you're not changing other fields, while keeping the code simple?

Also, I think there may still be a race condition whereby another process might set a new version, and then this process could set a lower version.

Process A:
* Increment `configuration_version` to 2 and `refresh_from_db()`

Process B:
* Increment `configuration_version` to 3 and `refresh_from_db()`
* Obtain `_configuration_lock`; at this point, `candidate_configuration_version` is 3
* In atomic transaction, set `deployed_configuration_version` to 3

Process A:
* Obtain `_configuration_lock`; at this point, `candidate_configuration_version` is 2
* In atomic transaction, set `deployed_configuration_version` to 2

Perhaps you should move the first `refresh_from_db()` within the lock and/or change the `with transaction.atomic()... save()` expression to:

```
LoadBalancingServer.objects.filter(pk=self.pk, deployed_configuration_version__lt=candidate_configuration_version).update(deployed_configuration_version=candidate_configuration_version)
```

Reply:
I think in a race condition situation, Process A wouldn't actually be able to acquire the lock in that scenario because we aren't blocking - it'd just get kicked out of the try/except block as soon as it tried to grab it because Process B already did. Or did I misunderstand something?

Reply:
I can change this to an `update()` call if you prefer, though I don't think the current code could change any other fields.  We refresh from the DB after starting the transaction, so the DB row will be locked.

And I don't think what you describe is a race condition.  First, I don't think it can happen, since we don't block when trying to grab the lock, so this would only happen when process A does not get a time slice for a full LB reconfiguration, which is for a minute or so.  And even if it does happen, the result would still be correct – we just would trigger a redundant reconfiguration, but I don't think it's even possible to avoid _any_ redundant reconfiguration while at the same time never missing one (in fact I think it's quite easy to prove that this is impossible).

So I can move the `refresh_from_db()` inside the `with` statement, but I think it's clearer as it is, and it does not make much of a difference.

Reply:
@UmanShahzad It's actually theroretically possible, since processes can be interrupted by the OS at any time, and aren't any guarantees at all for how long they will be interrupted.  In practice, however, and interruption of a minute (or even just several seconds) is rather unlikely unless the system is in a very unhealthy state (thrashing, ooming, lots of proccesses waiting for CPU, etc).

Reply:
@bradenmacdonald Sounds all reasonable, I made that change.  the downside is that we now have to wait at least two hours again before we can merge this.  I'll test on stage in the meantime.